### PR TITLE
introducing trie tree matching

### PR DESCRIPTION
**Pull request by wulianglongrd**
This PR attempts to implement [this comment's suggestion](https://github.com/istio/istio/pull/46043#discussion_r1265601483):

Use trie structure to match and abstract.

Currently it is just an initialization implementation and only applies to the `selectVirtualServices` and `selectServices` methods. Gather feedback from the community first before considering further application elsewhere (It can be used for host matching of `VS`, `DR` and `service` during `InitContext`).

### CODE REVIEW

File: pilot/pkg/model/virtualservice.go [hunk 1]
```diff
@@ -35,72 +35,47 @@ import (
...
 	loopAndAdd(vsidx.privateByNamespaceAndGateway[n])
 	loopAndAdd(vsidx.exportedToNamespaceByGateway[n])
 	loopAndAdd(vsidx.publicByGateway[constants.IstioMeshGateway])
 
-	return importedVirtualServices
-}
+	exists := sets.NewWithLength[string](len(importedVirtualServices))
+	slices.FilterInPlace(importedVirtualServices, func(vs config.Config) bool {
+		name := vs.Name + "/" + vs.Namespace
+		return !exists.InsertContains(name)
```
Comment (howardjohn) [inline pilot/pkg/model/virtualservice.go:75]: As noted in ohter PR, we should not be mutating state inside closures passed to `slices`; they are intended to be idempotent and stateless functions

**Review by howardjohn — COMMENTED** [review]
Its hard to evaluate this without context. What problem are we tying to solve, and how well does this solve the problem?

I assume this is a performance thing. How much aggregate CPU is spent on VS matching in your environment with and without this patch?

**Comment by wulianglongrd** [comment]
> Its hard to evaluate this without context. What problem are we tying to solve, and how well does this solve the problem?
> 
> I assume this is a performance thing. How much aggregate CPU is spent on VS matching in your environment with and without this patch?

Hello @howardjohn, I updated the CR description. Currently only benchmarks are used for simple evaluation and not yet used in production environments. The benchmark evaluation does not include the cost of building the trie tree, and I haven't thought of a way to do an overall evaluation.

**Comment by wulianglongrd** [comment]
According to the `BenchmarkInitPushContext` test evaluation results, the performance has not improved significantly, but has declined (building a trie tree is the main reason). 

Although I believe that if used in a production environment, the effect should be better than the benchmark results (because `updateContext` does not always need to build a trie tree in a production environment), but I feel that its improvement is not worth the complexity. So I will close this PR.

**Pull request closed without merging**